Add the append method for the Headers API

jeenalee opened the pull request:
<!-- Please describe your changes on the following line: -->

This commit adds the append method for the Headers API. @malisas and I are both contributors. 

There are a few TODOs related:
- The script needs to parse the header value for certain header names to decide the header group it belongs
- There are possible spec bugs that could change what a valid header value looks like (related: [issue page](https://github.com/whatwg/fetch/issues/332))

There are WPT tests already written for the Headers API, but they will fail as the Headers API is not fully implemented. 

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [X] These changes do not require tests because tests for the Headers API already exists, but this commit does not implement the interface fully. The tests will fail.

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/12467)

<!-- Reviewable:end -->

jeenalee committed:
Add the append method for the Headers API for the Fetch API

This commit will add the append method and associated helper functions, and introduce any necessary changes for it.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/webidls/Headers.webidl, components/script/dom/mod.rs, components/script/dom/headers.rs, components/script/dom/bindings/str.rs

fitzgen commented:
r? @jdm

jeenalee commented:
r? @jdm 
cc: @malisas

jdm commented:
Good start! I've left some comments on ways I think we can make the code more readable.
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 6 of 6 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 30 unresolved discussions, some commit checks failed.

---

*[components/script/dom/headers.rs, line 39 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk1NJaCOmXnAlqrlc5:-KMk1NJbwDVCc91Hy8t0:1810244727) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L39)):*
> ```Rust
> 
> impl Headers {
>     pub fn new() -> Headers {
> ```

This should be called `new_inherited`, and a `new` method should return `Root<Headers>`.

---

*[components/script/dom/headers.rs, line 61 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk1XPUjYs77kc3XA1e:-KMk1XPUjYs77kc3XA1f:354228139) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L61)):*
> ```Rust
>         //    and name/value is not a CORS-safelisted request-header, return.
>         // 6) Otherwise, if guard is "response" and
>         //    name is a forbidden response-header name, return.
> ```

Rather than writing out comments that duplicate specification text, we should use `// Steps 3-6` and let the reader look up the spec if necessary.

---

*[components/script/dom/headers.rs, line 66 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk1yNngppoO9T_k8Yk:-KMk1yNngppoO9T_k8Yl:1840832128) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L66)):*
> ```Rust
>             Ok(ns) => name_string = ns,
>             Err(_) => return Err(Error::Type(String::from("Non-UTF8 header name found"))),
>         }
> ```

```rust
let name_string = match String::from_utf8(name.into()) {
    Ok(ns) => ns,
    _ => return Err(Error::Type("Non-UTF8 header name found".into()));
};
````

---

*[components/script/dom/headers.rs, line 67 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMkCIDi1imBPJAp2qi4:-KMkCIDi1imBPJAp2qi5:-322200205) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L67)):*
> ```Rust
>             Err(_) => return Err(Error::Type(String::from("Non-UTF8 header name found"))),
>         }
>         let header_group = try!(find_header_group(&name_string));
> ```

I think this header group abstraction makes it harder to reason about what this code is actually checking. I'd prefer to have each step written out like in the spec.

---

*[components/script/dom/headers.rs, line 86 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMkAWQf5Ze4JDyKcR_a:-KMkAWQf5Ze4JDyKcR_b:-1655846712) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L86)):*
> ```Rust
> // "DPR", "Downlink", "Save-Data", "Viewport-Width", "Width":
> // once parsed, the value should not be failure.
> fn is_cors_safelisted_request(name: &str) -> bool {
> ```

`is_cors_safelisted_request_header` and link to https://fetch.spec.whatwg.org/#cors-safelisted-request-header .

---

*[components/script/dom/headers.rs, line 95 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMkAQ68jaNELcFXc0yX:-KMkAQ68jaNELcFXc0yY:-2088739518) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L95)):*
> ```Rust
> }
> 
> fn is_cors_non_wildcard_request(name: &str) -> bool {
> ```

This isn't used by append.

---

*[components/script/dom/headers.rs, line 102 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk9BxQ2uAfECx9oOj_:-KMk9BxR6GnmMsdJXvGi:69318877) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L102)):*
> ```Rust
> }
> 
> // TODO
> ```

What's left to do?

---

*[components/script/dom/headers.rs, line 108 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk9xG9cpDjKzIcoyM_:-KMk9E-I0O80A2WbYApv:1668030296) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L108)):*
> ```Rust
> // Content-Type, Expires, Last-Modified, Pragma
> // Any value in list that is not a forbidden response-header name.
> fn is_cors_safelisted_response(name: &str) -> bool {
> ```

What does this correspond to in the specification? Append doesn't appear to check for these.

---

*[components/script/dom/headers.rs, line 120 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMkABBd3xyEamPsfuP2:-KMkABBd3xyEamPsfuP3:603570704) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L120)):*
> ```Rust
> }
> 
> fn is_forbidden_response(name: &str) -> bool {
> ```

`is_forbidden_response_header` and add a link to https://fetch.spec.whatwg.org/#forbidden-response-header-name

---

*[components/script/dom/headers.rs, line 128 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMkAIIYddf9ozDKp7rS:-KMkAIIZziRYcGOXtjNS:21945740) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L128)):*
> ```Rust
> }
> 
> fn is_forbidden(name: &str) -> bool {
> ```

`is_forbidden_header_name` and add a link to https://fetch.spec.whatwg.org/#forbidden-header-name

---

*[components/script/dom/headers.rs, line 129 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk8SCz7KIVhXVPeU9L:-KMk8SD-WGENt3rvtNjf:-794012434) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L129)):*
> ```Rust
> 
> fn is_forbidden(name: &str) -> bool {
>     if name.starts_with("proxy-") ||
> ```

If we could merge this with https://dxr.mozilla.org/servo/rev/4cc4bc3c38eff22f9472215261d6926315b42f8d/components/script/dom/xmlhttprequest.rs#416 that would be great. The code there is shorter; I think it would be good to replace this implementation with that one, and have that one call into this one.

---

*[components/script/dom/headers.rs, line 198 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk3XFCRU6Jvz1YSssf:-KMk3XFCRU6Jvz1YSssg:288794753) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L198)):*
> ```Rust
> // [3] https://tools.ietf.org/html/rfc7230#section-3.2.6
> // [4] https://www.rfc-editor.org/errata_search.php?rfc=7230
> fn validate_name_and_value(name: &ByteString, value: &ByteString) -> Result<(), Error> {
> ```

Let's make this method return `Result<(String, Vec<u8>), Error>` so the caller only needs to deal with valid values after this point. We can accept ByteString by value instead of as a reference to accomplish this.

---

*[components/script/dom/headers.rs, line 202 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk3JPWz_1y0yGB6uIN:-KMk3JPXwwQdjSVW9hJR:1575663478) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L202)):*
> ```Rust
>         Ok(())
>     } else {
>         Err(Error::Type(String::from("Name and/or Value are not valid")))
> ```

We can have a more specific error message if we do:
```rust
if !is_field_name(name) {
    return Err(...);
}
if !is_field_content(value) {
    return Err(...);
}
Ok(())
```

---

*[components/script/dom/headers.rs, line 220 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMkD43cIZ7L_9cOkcW7:-KMkD43d2nByI7SsrwyN:883981979) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L220)):*
> ```Rust
> 
> // Removes trailing and leading HTTP whitespace bytes.
> pub fn normalize_value(value: ByteString) -> ByteString {
> ```

Add a comment linking to https://fetch.spec.whatwg.org/#concept-header-value-normalize

---

*[components/script/dom/headers.rs, line 222 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMkDFkz7CSKbHsWhGR3:-KMkDFkz7CSKbHsWhGR4:1759401572) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L222)):*
> ```Rust
> pub fn normalize_value(value: ByteString) -> ByteString {
>     let opt_first_index = index_of_first_non_whitespace(&value);
>     match opt_first_index {
> ```

What about this instead?
```rust
match (index_of_first_non_whitespace(&value), index_of_last_non_whitespace(&value)) {
    (Some(begin), Some(end)) => ByteString::new(value[begin..end].to_owned()),
    _ => ByteString::new(),
}
```

---

*[components/script/dom/headers.rs, line 255 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk42DwEQDYpuoQZJ2H:-KMk42DxCGmud3QKHwNa:-2002986909) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L255)):*
> ```Rust
>         byte == 0x0A ||      // new line
>         byte == 0x0D ||      // return character
>         byte == 0x20;        // space
> ```

This can be written as:
```rust
fn is_HTTP_whitespace(byte: u8) -> bool {
    byte == b'\t' || byte == b'\n' || byte == b'\r' || b == ' '
}
```

---

*[components/script/dom/headers.rs, line 263 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk4sUGewr8iiL33_bC:-KMk4sUHtjteRhVgjvBi:1266547771) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L263)):*
> ```Rust
>             continue;
>         }
>         return Some(index)
> ```

Let's write this as:
```rust
if !is_HTTP_whitespace(byte) {
    return Some(index);
}
```

---

*[components/script/dom/headers.rs, line 273 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk51-A0KShrj2AOtZm:-KMk51-A0KShrj2AOtZn:-482910493) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L273)):*
> ```Rust
>             continue;
>         }
>         return Some(index)
> ```

Same previous comment.

---

*[components/script/dom/headers.rs, line 284 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk5TpY_cK30A0SepMr:-KMk5TpY_cK30A0SepMs:-186229613) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L284)):*
> ```Rust
> 
> fn is_field_content(value: &ByteString) -> bool {
>     // http://tools.ietf.org/html/rfc2616#section-2.2
> ```

https://tools.ietf.org/html/rfc7230#section-3.2

---

*[components/script/dom/headers.rs, line 310 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk58AVXcAttF5t5i9S:-KMk58AWZq1-b0IfWv4K:-1752249469) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L310)):*
> ```Rust
> 
> fn is_space(x: u8) -> bool {
>     return x == 0x20;
> ```

This can be `x == b' '` without the `return`.

---

*[components/script/dom/headers.rs, line 314 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk5E-OceSFAODvYmo4:-KMk5E-OceSFAODvYmo5:-634548432) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L314)):*
> ```Rust
> 
> fn is_htab(x: u8) -> bool {
>     return x == 0x09;
> ```

This can be `x == b'\t'`.

---

*[components/script/dom/headers.rs, line 322 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk5R2uALcTwauJlpF8:-KMk5R2uALcTwauJlpF9:-186229613) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L322)):*
> ```Rust
> 
> fn is_vchar(x: u8) -> bool {
>     // http://tools.ietf.org/html/rfc2616#section-2.2
> ```

https://tools.ietf.org/html/rfc7230#section-3.2

---

*[components/script/dom/headers.rs, line 325 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk5cki6rtFam5mOqnG:-KMk5ckj9fGl6jhdMN0x:1821011370) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L325)):*
> ```Rust
>     // field-vchar = VCHAR / obs-text
>     match x {
>         0...31 | 127 => false, // CTLs
> ```

Let's use `0x00...0x1F | 0x7F` per https://tools.ietf.org/html/rfc5234#appendix-B.1, but I'm not clear on where CTLs come into the field-vchar definition?

---

*[components/script/dom/headers.rs, line 326 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk6pltGCEUGj7nw-9q:-KMk6pltGCEUGj7nw-9r:-465550562) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L326)):*
> ```Rust
>     match x {
>         0...31 | 127 => false, // CTLs
>         x if x > 127 => false, // non-CHARs
> ```

Could we phrase the checks in this function more like `0x21...0x7E` in https://tools.ietf.org/html/rfc5234#appendix-B.1

---

*[components/script/dom/headers.rs, line 327 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk6GNEX3R8xRY7JFb_:-KMk6GNFmR0KefApuOhQ:-337971033) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L327)):*
> ```Rust
>         0...31 | 127 => false, // CTLs
>         x if x > 127 => false, // non-CHARs
>         _ => true,
> ```

Can we reuse is_obs_text here?

---

*[components/script/dom/webidls/Headers.webidl, line 5 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk73nVt9onb12z6xn2:-KMk73nVt9onb12z6xn3:1392552296) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/webidls/Headers.webidl#L5)):*
> ```WebIDL
>  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
> 
> // https://fetch.spec.whatwg.org/#header
> ```

https://fetch.spec.whatwg.org/#headers-class

---

*[tests/unit/script/headers.rs, line 14 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk7AI8gBTBdyyIXDvH:-KMk7AI8gBTBdyyIXDvI:871948107) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/tests/unit/script/headers.rs#L14)):*
> ```Rust
>     let actual_normalized_empty_bytestring = headers::normalize(empty_bytestring);
>     let expected_normalized_empty_bytestring = ByteString::new(vec![]);
>     assert_eq!(actual_normalized_empty_bytestring, expected_normalized_empty_bytestring);
> ```

It will be easier to read if these are called `actual` and `expected` instead.

---

*[tests/unit/script/headers.rs, line 20 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk7GGjXb08eHkORuCT:-KMk7GGjXb08eHkORuCU:101002505) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/tests/unit/script/headers.rs#L20)):*
> ```Rust
> fn test_normalize_all_whitespace_bytestring() {
>     // All whitespace test. A horizontal tab, a line feed, a carriage return , and a space
>     let all_whitespace_bytestring = ByteString::new(vec![0x09, 0x0A, '\n' as u8, 0x20]);
> ```

`[b'\t', b'\n', b'\r', b' ']`

---

*[tests/unit/script/headers.rs, line 23 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk7Njf59PWi-_O9-XQ:-KMk7Njf59PWi-_O9-XR:-807610634) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/tests/unit/script/headers.rs#L23)):*
> ```Rust
>     let actual_normalized_whitespace_bytestring = headers::normalize(all_whitespace_bytestring);
>     let expected_normalized_whitespace_bytestring = ByteString::new(vec![]);
>     assert_eq!(actual_normalized_whitespace_bytestring, expected_normalized_whitespace_bytestring);
> ```

Same naming comment for all the tests.

---

*[tests/unit/script/headers.rs, line 29 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk7xZginmvjyIM5yNh:-KMk7RTx7vcVCLKS83dL:2078169649) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/tests/unit/script/headers.rs#L29)):*
> ```Rust
> fn test_normalize_non_empty_no_whitespace_bytestring() {
>     // Non-empty, no whitespace ByteString test
>     let no_whitespace_bytestring = ByteString::new(vec!['S' as u8, '!' as u8]);
> ```

We can use `b` prefixes to avoid the casts.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12467#-:-KMkErOwXeKtVrBDiHMZ:-872077430)*
<!-- Sent from Reviewable.io -->

nox commented:
AFAIK the spec links should be outside the functions, not inside.

malisas commented:
@jdm 
@jeenalee and I went through your comments and made some changes. The tests also pass now after rebasing. Please let us know if we should change anything else before merging -- thanks! :)

jdm commented:
Thanks for addressing those comments! There are still some unaddressed ones in the unit tests, and I've left some new ones.
-S-awaiting-review +S-needs-code-changes 

---

Reviewed 1 of 1 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[components/script/dom/headers.rs, line 62 \[r5\]](https://reviewable.io:443/reviews/servo/servo/12467#-KN3AQJVTr2Cz0vxxNQ2:-KN3AQJWVFb-ZTZso4Yu:-411080248) ([raw file](https://github.com/servo/servo/blob/3d477f9d4aa48db06f2db8f969d3214e2757e729/components/script/dom/headers.rs#L62)):*
> ```Rust
>         // Step 2
>         let (valid_name, valid_value) = try!(validate_name_and_value(name, value));
>         let header_group = try!(find_header_group(&valid_name));
> ```

I still find the following code more readable than the header group abstraction:
```rust
// Step 3
if self.guard == Guard::Immutable {
    return ...;
}

// Step 4
if self.guard == Guard::Request && is_forbidden_header_name(valid_name) {
    return ...;
}
```

---

*[components/script/dom/headers.rs, line 123 \[r5\]](https://reviewable.io:443/reviews/servo/servo/12467#-KN3BUfxgEJXAGM0O2Ai:-KN3BUfxgEJXAGM0O2Aj:-1991735675) ([raw file](https://github.com/servo/servo/blob/3d477f9d4aa48db06f2db8f969d3214e2757e729/components/script/dom/headers.rs#L123)):*
> ```Rust
>     let disallowed_header_prefixes = ["sec-", "proxy-"];
> 
>     if disallowed_headers.iter().any(|header| *header == name) ||
> ```

Instead of `if ... { return true } else { return false }`, we can write `...` which is equivalent.

---

*[components/script/dom/headers.rs, line 176 \[r5\]](https://reviewable.io:443/reviews/servo/servo/12467#-KN3BlnbtmUZHuhEcdRw:-KN3BlnbtmUZHuhEcdRx:155851381) ([raw file](https://github.com/servo/servo/blob/3d477f9d4aa48db06f2db8f969d3214e2757e729/components/script/dom/headers.rs#L176)):*
> ```Rust
>     match String::from_utf8(name.into()) {
>         Ok(ns) => Ok((ns, value.into())),
>         _ => return Err(Error::Type("Non-UTF8 header name found".to_string())),
> ```

No need for `return` here.

---

*[components/script/dom/headers.rs, line 205 \[r5\]](https://reviewable.io:443/reviews/servo/servo/12467#-KN3C5VeXmfkEbAbgK6n:-KN3C5VeXmfkEbAbgK6o:66228605) ([raw file](https://github.com/servo/servo/blob/3d477f9d4aa48db06f2db8f969d3214e2757e729/components/script/dom/headers.rs#L205)):*
> ```Rust
>     for (index, &byte) in value.iter().enumerate().rev() {
>         if !is_HTTP_whitespace(byte) {
>             return Some(index);
> ```

We should return `Some(value.length() - index - 1)` instead, or it will be easy to use the value incorrectly like we did earlier.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12467#-:-KN3CoSqxCH08vg7yYXx:605170723)*
<!-- Sent from Reviewable.io -->

jeenalee commented:
*[components/script/dom/headers.rs, line 205 \[r5\]](https://reviewable.io:443/reviews/servo/servo/12467#-KN3C5VeXmfkEbAbgK6n:-KN3TxSZDWJl0ckaJSa7:1550619194) ([raw file](https://github.com/servo/servo/blob/3d477f9d4aa48db06f2db8f969d3214e2757e729/components/script/dom/headers.rs#L205)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> We should return `Some(value.length() - index - 1)` instead, or it will be easy to use the value incorrectly like we did earlier.
</details>
I think it will have to return what you suggested if it were `value.iter().rev().enumerate()`. But because it is written as `enumerate()` then `rev()`, I think it returns the index in the form we want. What do you think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12467)*
<!-- Sent from Reviewable.io -->

Manishearth commented on components/script/dom/headers.rs:
```diff
@@ -0,0 +1,269 @@
+
+fn index_of_last_non_whitespace(value: &ByteString) -> Option<usize> {
+    for (index, &byte) in value.iter().enumerate().rev() {
+        if !is_HTTP_whitespace(byte) {
```
We can use even more iterator magic here!

`value.iter().enumerate().rev().find(|_,v| !is_whitespace(v)).map(|i| i.0|)`

something like that.

But it's good as is too, some folks prefer to avoid overusing iterator adapters.

jdm commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[components/script/dom/headers.rs, line 205 \[r5\]](https://reviewable.io:443/reviews/servo/servo/12467#-KN3C5VeXmfkEbAbgK6n:-KN3ZjubzECAJDfBLJK8:-1196771673) ([raw file](https://github.com/servo/servo/blob/3d477f9d4aa48db06f2db8f969d3214e2757e729/components/script/dom/headers.rs#L205)):*
<details><summary><i>Previously, jeenalee (Jeena Lee) wrote…</i></summary>
> I think it will have to return what you suggested if it were `value.iter().rev().enumerate()`. But because it is written as `enumerate()` then `rev()`, I think it returns the index in the form we want. What do you think?
</details>
Ooh, you're right. That's subtle!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12467)*
<!-- Sent from Reviewable.io -->

jdm commented:
Almost there! Please go ahead and [squash](https://ariejan.net/2011/07/05/git-squash-your-latests-commits-into-one/) all the changes together into a single commit, as well. I don't think we gain much from having the independent commits when they're all part of implementing a single feature.
-S-awaiting-review +S-needs-code-changes +S-needs-squash 

---

Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[components/script/dom/headers.rs, line 129 \[r2\]](https://reviewable.io:443/reviews/servo/servo/12467#-KMk8SCz7KIVhXVPeU9L:-KN3eU6RE0Ufo8sSwoJ_:71600666) ([raw file](https://github.com/servo/servo/blob/e40516194fe4b9fd2f51b32026d3f4505482fe9c/components/script/dom/headers.rs#L129)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary>
> If we could merge this with https://dxr.mozilla.org/servo/rev/4cc4bc3c38eff22f9472215261d6926315b42f8d/components/script/dom/xmlhttprequest.rs#416 that would be great. The code there is shorter; I think it would be good to replace this implementation with that one, and have that one call into this one.
</details>
What about calling this from xmlhttprequest.rs?

---

*[components/script/dom/headers.rs, line 60 \[r6\]](https://reviewable.io:443/reviews/servo/servo/12467#-KN3dy54nmnIUUNn_5hy:-KN3dy55b7zd1If7xK6o:-1084559919) ([raw file](https://github.com/servo/servo/blob/8f4d953774718609b960ab705d691b1008d1a6c5/components/script/dom/headers.rs#L60)):*
> ```Rust
>         // Step 4
>         if self.guard == Guard::Request && is_forbidden_header_name(&valid_name) {
>             return Ok(());;
> ```

nit: extra ;

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12467#-:-KN3eXSlkyGkIH2t9ogH:1421821156)*
<!-- Sent from Reviewable.io -->

malisas committed:
makes XMLHttpRequest::SetRequestHeader call dom::headers::is_forbidden_header_name

jdm commented:
@bors-servo: r+
Great work!

---

Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/12467#-:-KN403PSLM5KHLAbE5A7:171583111)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit 60549c4 has been approved by `jdm`

<!-- @bors-servo r=jdm 60549c47ba4383784b2b11ab8805099f1ed1a241 -->

bors-servo commented:
:hourglass: Testing commit 60549c4 with merge 0bcf291...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/2236)

jdm commented:
```
Tests with unexpected results:
  ▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.html:
  │ FAIL [expected PASS] Interfaces exposed on the window
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface Headers to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false
  │ 
  │ test_interfaces/<@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:70:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1402:20
  │ test@http://web-platform.test:8000/resources/testharness.js:500:9
  │ test_interfaces@http://web-platform.test:8000/_mozilla/mozilla/interfaces.js:1:3
  └ @http://web-platform.test:8000/_mozilla/mozilla/interfaces.html:6:1
```
We'll need to add the Headers interface name to the list in `tests/wpt/mozilla/tests/mozilla/interfaces.js`.

jeenalee committed:
List the Headers interface name to the list in interfaces test files.

jdm commented on tests/wpt/mozilla/tests/mozilla/interfaces.worker.js:
```diff
@@ -44,6 +44,7 @@ test_interfaces([
   "FocusEvent",
   "FormData",
   "HashChangeEvent",
+  "Headers",  
```
This won't work yet, since the `Exposed=(Window,Worker)` annotation is commented out.

malisas committed:
Small changes to Headers.webidl and headers.rs

Expose the Headers interface to the Window scope and to Workers
Move Append function inside HeadersMethods trait

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 094995b has been approved by `jdm`

<!-- @bors-servo r=jdm 094995b9cbb0f51d26e59ebf71cfed31a4458380 -->

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 272d12d has been approved by `jdm`

<!-- @bors-servo r=jdm 272d12da26f150f9fd6447572dc10464340f3da4 -->

bors-servo commented:
:hourglass: Testing commit 272d12d with merge 03fa7f0...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1975), [arm32](http://build.servo.org/builders/arm32/builds/2088), [arm64](http://build.servo.org/builders/arm64/builds/2089), [linux-dev](http://build.servo.org/builders/linux-dev/builds/2132), [linux-rel](http://build.servo.org/builders/linux-rel/builds/2252), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/2161), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/2148), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/2220), [windows-dev](http://build.servo.org/builders/windows-dev/builds/34)
